src/app/application/services/data-source/index.ts

```diff
@@ -31,6 +33,37 @@ export const defineSources = <T extends Record<PropertyKey, unknown>>(sources: E
   return sources
 }
 
+type Sources = Record<PropertyKey, (...args: any[]) => any>
+type SourceFactory = (...args: any[]) => Sources
+export type TypeOf<T> = T extends { typeOf(): any } ? ReturnType<T['typeOf']> : any
+
+class TypedString<T = unknown> {
+  constructor(
+    protected str: string,
+  ) {}
+
+  toString() {
+    return this.str
+  }
+
+  typeOf() {
+    return undefined as T
+  }
+}
+
+const cache = new Map()
+export const useUri = () => {
+  return <T extends SourceFactory, K extends keyof ReturnType<T>>(_typeRef: T, src: K, params: ExtractRouteParams<K>, query: Partial<PaginationParams> = {}) => {
+    const str = String(src)
+    if (!cache.has(str)) {
+      cache.set(str, compile(str))
+    }
+    const interpolate = cache.get(str)
+    const q = Object.entries(query).reduce<string[]>((prev, [key, value]) => prev.concat(`${key}=${value}`), [])
+    const uri = `${interpolate(params)}${q.length > 0 ? `?${q.join('&')}` : ''}`
```

Comment:
The above few lines are the only real meaningful runtime additions in this PR

1. Compile the path/uri/route (`/this/:thing`) into something you can interpolate with the variables using the `path-to-regexp` util
2. Add this to a cache so we only do this once per URI.
3. interpolate the route with the parameters
4. Potentially make a query string to add on the end.

The rest is just for Typescript